wallet/datastore.go

```diff
@@ -0,0 +1,333 @@
+			return nil, nil, err
+		}
+	}
+	return walletDB, walletRODB, nil
+}
+
+var (
+	// ErrTooManyCardsLinked denotes when more than 3 cards have been linked to a single wallet
+	ErrTooManyCardsLinked = errors.New("unable to add too many wallets to a single user")
+)
+
+// UpsertWallet upserts the given wallet
+func (pg *Postgres) UpsertWallet(wallet *wallet.Info) error {
+	statement := `
+	insert into wallets (id, provider, provider_id, public_key, provider_linking_id, anonymous_address)
+	values ($1, $2, $3, $4, $5, $6)
+	on conflict (id) do
+	update set
+		provider = $2,
+		provider_id = $3,
+		provider_linking_id = $5,
+		anonymous_address = $6
+	returning *`
+	_, err := pg.RawDB().Exec(statement, wallet.ID, wallet.Provider, wallet.ProviderID, wallet.PublicKey, wallet.ProviderLinkingID, wallet.AnonymousAddress)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetWallet by ID
+func (pg *Postgres) GetWallet(ID uuid.UUID) (*wallet.Info, error) {
+	statement := "select * from wallets where id = $1"
+	wallets := []wallet.Info{}
+	err := pg.RawDB().Select(&wallets, statement, ID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(wallets) > 0 {
+		// FIXME currently assumes BAT
+		{
+			tmp := altcurrency.BAT
+			wallets[0].AltCurrency = &tmp
+		}
+		return &wallets[0], nil
+	}
+
+	return nil, nil
+}
+
+// GetWalletByPublicKey gets a wallet by a public key
+func (pg *Postgres) GetWalletByPublicKey(pk string) (*walletutils.Info, error) {
+	statement := `
+	SELECT *
+	FROM wallets
+	WHERE public_key = $1
+	`
+	var wallet walletutils.Info
+	err := pg.RawDB().Get(&wallet, statement, pk)
+	return &wallet, err
+}
+
+// GetByProviderLinkingID gets a wallet by a provider address
+func (pg *Postgres) GetByProviderLinkingID(providerLinkingID uuid.UUID) (*[]walletutils.Info, error) {
+	statement := `
+	SELECT *
+	FROM wallets
+	WHERE provider_linking_id = $1
+	`
+	var wallets []walletutils.Info
+	err := pg.RawDB().Select(&wallets, statement, providerLinkingID)
+	return &wallets, err
+}
+
+// InsertWallet inserts the given wallet
+func (pg *Postgres) InsertWallet(wallet *walletutils.Info) error {
+	// NOTE on conflict do nothing because none of the wallet information is updateable
+	statement := `
+	INSERT INTO wallets (id, provider, provider_id, public_key)
+	VALUES ($1, $2, $3, $4)
+	ON CONFLICT DO NOTHING`
+	_, err := pg.RawDB().Exec(
+		statement,
+		wallet.ID,
+		wallet.Provider,
+		wallet.ProviderID,
+		wallet.PublicKey,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetAnonymousAddress sets the anon addresses of the provided wallets
+func (pg *Postgres) SetAnonymousAddress(ID string, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			anonymous_address = $2
+	WHERE id = $1
+	`
+	_, err := pg.RawDB().Exec(
+		statement,
+		ID,
+		anonymousAddress,
+	)
+	return err
+}
+
+// LinkWallet links a wallet and transfers funds to newly linked wallet
+func (service *Service) LinkWallet(
```

Comment:
Can we move this function into the Service struct file? It doesn't seem like this would belong in here since it's calling `service.Datastore`?

Reply:
:+1:

Reply:
https://github.com/brave-intl/bat-go/pull/314/commits/d9ad61f5ecbbf6e774a35d6481720162bacc3f87